### PR TITLE
[mlir][SPIR-V] Update the `ConvertToSPIRV` pass to use dialect interfaces

### PR DESCRIPTION
**Pull request by fabianmcg**
This patch updates the base implementation of `ConvertToSPIRV` to be more like the implementation of `ConvertToLLVM`. `ConvertToLLVM` relies on dialect interfaces for configuring the conversion, allowing out-of-tree dialects to participate in the pass if they implement the interface.

This patch introduces the `ConvertToSPIRVPatternInterface` dialect interface, allowing the configuration of the conversion to SPIR-V on a dialect per dialect basis.

Finally, this patch adds the dialect interfaces for all previously supported dialects in the previous implementation of the `ConvertToSPIRV` pass.

Note:
The convert SCF to SPIR-V was left inside the pass, as it depends on the `ScfToSPIRVContext`, a TODO for a future patch is removing this issue.

### CODE REVIEW

File: mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp [hunk 1]
```diff
@@ -1367,3 +1369,39 @@ struct ConvertArithToSPIRVPass
...
 }
+
+//===----------------------------------------------------------------------===//
+// ConvertToSPIRVPatternInterface implementation
+//===----------------------------------------------------------------------===//
+namespace {
+/// Implement the interface to convert arith to SPIR-V.
+struct ToSPIRVDialectInterface : public ConvertToSPIRVPatternInterface {
+  using ConvertToSPIRVPatternInterface::ConvertToSPIRVPatternInterface;
+  void loadDependentDialects(MLIRContext *context) const final {
```
Comment (Hardcode84) [inline mlir/lib/Conversion/ArithToSPIRV/ArithToSPIRV.cpp:1380]: Here and later: `override`

File: mlir/lib/Conversion/ConvertToSPIRV/ToSPIRVInterface.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/Operation.h"
+#include "llvm/ADT/DenseSet.h"
+
+using namespace mlir;
+
+void mlir::populateConversionTargetFromOperation(
+    Operation *root, ConversionTarget &target,
+    SPIRVTypeConverter &typeConverter, RewritePatternSet &patterns) {
+  DenseSet<Dialect *> dialects;
```
Comment (Hardcode84) [inline mlir/lib/Conversion/ConvertToSPIRV/ToSPIRVInterface.cpp:19]: `SmallDenseSet`

File: mlir/lib/Conversion/IndexToSPIRV/IndexToSPIRV.cpp [hunk 3]
```diff
@@ -416,3 +417,37 @@ struct ConvertIndexToSPIRVPass
...
+  }
+
+  /// Hook for derived dialect interface to provide conversion patterns
+  /// and mark dialect legal for the conversion target.
+  void populateConvertToSPIRVConversionPatterns(
+      ConversionTarget &target, SPIRVTypeConverter &typeConverter,
+      RewritePatternSet &patterns) const final {
+    // Use UnrealizedConversionCast as the bridge so that we don't need to pull
+    // in patterns for other dialects.
+    target.addLegalOp<UnrealizedConversionCastOp>();
```
Comment (Hardcode84) [inline mlir/lib/Conversion/IndexToSPIRV/IndexToSPIRV.cpp:439]: This `addLegalOp<UnrealizedConversionCastOp>()` is copypasted to all interface implementations, maybe it should be handled by the pass itself?

**Review by kuhar — COMMENTED** [review]
This is interesting! What's your motivation for this beyond what's explained in the PR description? Is this to better align with the llvm conversion or do you have some out of tree dialects that you would like to support?

The reason I'm asking is that converting to SPIR-V has a set of issues that are unique to it and make it harder (I think?) than conversion to llvm:
* It does not support arbitrary (1d) vector types
* It does not support arbitrary integer types
* I think arrays are also more limited
* Not all ops are converted directly and some require expansion / emulation before conversion (e.g., masked vector load / store).

Because of these complications, our plan immediate was to stick with handcrafted conversion that happens in a few phases.

File: mlir/lib/Conversion/IndexToSPIRV/IndexToSPIRV.cpp (hunk 3 above)
Comment (fabianmcg) [reply mlir/lib/Conversion/IndexToSPIRV/IndexToSPIRV.cpp]: I removed it as it was not really needed when all patterns run together.

**Comment by fabianmcg** [comment]
> This is interesting! What's your motivation for this beyond what's explained in the PR description? Is this to better align with the llvm conversion or do you have some out of tree dialects that you would like to support?

It's both, I think it should align with `convert-to-llvm` and I have downstream project that needs to go to SPIR-V.

> The reason I'm asking is that converting to SPIR-V has a set of issues that are unique to it and make it harder (I think?) than conversion to llvm:
> 
> * It does not support arbitrary (1d) vector types
> * It does not support arbitrary integer types
> * I think arrays are also more limited
> * Not all ops are converted directly and some require expansion / emulation before conversion (e.g., masked vector load / store).
> 
> Because of these complications, our plan immediate was to stick with handcrafted conversion that happens in a few phases.

I noticed the vector issue with `cf-to-spirv`. If `convert-to-spirv` is run with unrolling activated, then it produces incorrect results.

I think some of the issues you mention can be solved within the dialect conversion infrastructure., or with missing intermediate ops.

**Comment by fabianmcg** [comment]
After talking with @kuhar we decided it's best to close this PR as the conversion to SPIR-V still needs to mature more before productivizing this pass.

**Pull request closed without merging**